ouroboros-network/src/Ouroboros/Network/Subscription/Common.hs

```diff
@@ -202,16 +202,28 @@ subscribeTo tbl threadPool tracer localIPv4_m localIPv6_m connectionAttemptDelay
                         else return ConnectFail
         traceWith tracer $ SubscriptionTraceConnectEnd remoteAddr result
         case result of
-            ConnectSuccess -> k sd -- TODO catch all application exception.
+            ConnectSuccess -> runApplication sd
             ConnectSuccessLast -> do
                     outstandingConThreads <- atomically $ readTVar conThreads
                     mapM_ (\a -> throwTo a
                             (SubscriberError SubscriberParrallelConnectionCancelled
                             "Parrallel connection cancelled"
                             callStack)) outstandingConThreads
-                    k sd
+                    runApplication sd
             ConnectFail -> return ()
 
+    runApplication :: Socket.Socket -> IO ()
+    runApplication sd = do
+        r_e <- try (k sd)
+        case r_e of
+             Left e -> do
+                 traceWith tracer $ SubscriptionApplicationException e
+                 -- TODO: Fatal/Permanent errors such as filesystem corruption should be
+                 -- permitted to trickle up instead of just causing the creation of a new
+                 -- connection.
+                 throwM e
```

Comment:
Idae: we need an extensible way to let some exception types declare that they are *fatal*, and use that here to check whether the given `SomeException` is fatal, in which case it should bubble up.

Reply:
Agreed.

Reply:
We can use the exceptions heirarchy mechanism to distinguish exceptions in particular categories.

See the docs for the pattern here https://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Exception.html#t:Exception